### PR TITLE
[ENG-505] feat: Add clerk login

### PR DESCRIPTION
**Pull request by eberle1080**
This PR makes the login and logout commands visible. It also refactors the clerk code so it can be used in a few different places. Finally, it adds support for clerk tokens for our backend calls.

### CODE REVIEW

**Review by laurenzlong — CHANGES_REQUESTED** [review]
A few issues:
- for the deploy command, the check for the existence of the key flag should be removed, if the key flag is not present then the Clerk token should be used. If there isn't a valid Clerk token, then an error should be printed to instruct the use to either run `amp login` or to provide a key flag

- I checked out this branch and ran `task build-prod`, then the login command (note: `lamp` is my alias for the local version of the CLI)
Screen Shot 2023-12-18 at 11 40 22 AM

I then cleared out my `~/Library/Application Support/Ampersand` folder and re-ran the command, got the same issue.

File: request/api.go [hunk 1]
```diff
@@ -100,19 +95,36 @@ func (c *APIClient) GetPreSignedUploadURL(ctx context.Context, md5 string) (Sign
...
-		// TODO: Default to token authentication and set Authorization header, instead of failing.
-		logger.Fatal("Must provide an API key in the --key flag")
+
+		return header, nil
 	}
 
-	return nil
+	jwt, err := clerk.FetchJwt(ctx)
+	if err != nil {
+		return Header{}, err
```
Comment (laurenzlong) [inline request/api.go:123]: Should call `logger.Fatalf` with an error instructing the user to either run `amp login` or to provide a key flag to the command.

File: request/api.go [hunk 2]
```diff
@@ -100,19 +95,47 @@ func (c *APIClient) GetPreSignedUploadURL(ctx context.Context, md5 string) (Sign
...
-		// TODO: Default to token authentication and set Authorization header, instead of failing.
-		logger.Fatal("Must provide an API key in the --key flag")
+		return Header{
+			Key:   "Authorization",
+			Value: fmt.Sprintf("Bearer %s", jwt),
+		}, nil
 	}
 
-	return nil
+	logger.Fatal("no authentication method found, please either log in (amp login) or provide an API key (--key)")
```
Comment (laurenzlong) [inline request/api.go:138]:
```suggestion
	logger.Fatal("no authentication method found, please run 'amp login' or provide an API key through the key flag")
```

**Review by laurenzlong — APPROVED** [review]
Can you also remove this: https://github.com/amp-labs/cli/blob/f9e18b8bc3ad1169a22cdc1f1c603ca7075376d3/cmd/deploy.go#L32

And then double check that you can run `amp login` without an API key and it successfully deploys?

**Comment by eberle1080** [comment]
> Can you also remove this:
> 
> https://github.com/amp-labs/cli/blob/f9e18b8bc3ad1169a22cdc1f1c603ca7075376d3/cmd/deploy.go#L32
> 
> And then double check that you can run `amp login` without an API key and it successfully deploys?
> 
> https://github.com/amp-labs/testdata has example amp.yaml files you can use for deployment.

I did as you instructed, everything is working :) 

> ../bin/amp deploy -p 20c23d64-99b6-495a-8ac7-c7ae11bd91fe .
> Successfully deployed your integration readAndWriteSalesforce.

**Pull request merged**